### ISSUE TITLE
why ONNX quantization works better on VIT?

### ISSUE THREAD

**Issue by AlexMercer-feng** [NVIDIA/TensorRT-Model-Optimizer#52]
I found a statement which said **3. Better support for vision transformers.** in link https://nvidia.github.io/TensorRT-Model-Optimizer/guides/_onnx_quantization.html.
I'm working on quantizing VIT now but can't find more infomation about the **better support**,could you give more explanation of why ONNX quantization works better than pytorch_quantization? thanks~

**Comment by i-riyad** [comment]
There are certain graph patterns that TensorRT can fuse together. `modelopt.onnx.quantization` takes those into consideration while quantizing certain layers. `pytorch_quantization` does not have those type of TensorRT backend specific analysis. But note that, not every ViT architecture is supported yet in `modelopt.onnx.quantization`.

**Comment by AlexMercer-feng** [comment]
> There are certain graph patterns that TensorRT can fuse together. `modelopt.onnx.quantization` takes those into consideration while quantizing certain layers. `pytorch_quantization` does not have those type of TensorRT backend specific analysis. But note that, not every ViT architecture is supported yet in `modelopt.onnx.quantization`.

I got it, thank you so much for your response. Btw, in my experience, not every Attention in different VIT architectures could run on MHA kernel after fused by Myelin, is the **unsupported yet** you mentioned also due to something like that? caused by TRT backend?

**Comment by hawl666** [comment]
hello, @i-riyad 
I use modelopt.onnx.quantization quantize my yolov8 model, get the qdq onnx, and use trtexec compile to the trt engine, finally the accuracy is poor.How can i fix this?

**Comment by i-riyad** [comment]
> I got it, thank you so much for your response. Btw, in my experience, not every Attention in different VIT architectures could run on MHA kernel after fused by Myelin, is the **unsupported yet** you mentioned also due to something like that? caused by TRT backend?

Thats right. Also, sometimes it can be caused my `modelopt` as well. For example, it is possible that fp16 MHA kernels were used by TRT for some non-quantized model/architecture but after explicit quantization with `modelopt` TRT might not find int8 MHA kernel for them and run using native int8 MatMuls.

**Comment by i-riyad** [comment]
> hello, @i-riyad I use modelopt.onnx.quantization quantize my yolov8 model, get the qdq onnx, and use trtexec compile to the trt engine, finally the accuracy is poor.How can i fix this?

How much is the regression? Have you used real data to calibrate the model?

**Comment by hawl666** [comment]
> > hello, @i-riyad I use modelopt.onnx.quantization quantize my yolov8 model, get the qdq onnx, and use trtexec compile to the trt engine, finally the accuracy is poor.How can i fix this?
> 
> How much is the regression? Have you used real data to calibrate the model?

yes, i used calibration data, and F1 score droped 50%

**Comment by i-riyad** [comment]
> yes, i used calibration data, and F1 score droped 50%

If the input ONNX model has trained weights, definitely there is something wrong with the calibration process. In our examples we have used a ViT model from `timm` and it requires [preprocessing of images](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/onnx_ptq/evaluate_vit.py#L62) before feeding into the model. Check if any kinda preprocessing is needed for your input model. You can also try to infer the model with the calibration dataset and see if it produces the expected outputs.

**Comment by hawl666** [comment]
> > yes, i used calibration data, and F1 score droped 50%
> 
> If the input ONNX model has trained weights, definitely there is something wrong with the calibration process. In our examples we have used a ViT model from `timm` and it requires [preprocessing of images](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/onnx_ptq/evaluate_vit.py#L62) before feeding into the model. Check if any kinda preprocessing is needed for your input model. You can also try to infer the model with the calibration dataset and see if it produces the expected outputs.

OK, I'll try this. Also, I used trtexec 8.4 to build the qdq onnx to engine on orin, is this correct?

**Comment by i-riyad** [comment]
> trtexec 8.4 to build the qdq onnx to engine on orin, is this correct?

8.4 should work but I would encourage to use [10.x](https://forums.developer.nvidia.com/t/is-it-possible-to-already-use-tensorrt-10-on-jestson-agx-orin/295744/9), specially for ViT models.

**Comment by hawl666** [comment]
> > > yes, i used calibration data, and F1 score droped 50%
> > 
> > 
> > If the input ONNX model has trained weights, definitely there is something wrong with the calibration process. In our examples we have used a ViT model from `timm` and it requires [preprocessing of images](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/onnx_ptq/evaluate_vit.py#L62) before feeding into the model. Check if any kinda preprocessing is needed for your input model. You can also try to infer the model with the calibration dataset and see if it produces the expected outputs.
> 
> OK, I'll try this. Also, I used trtexec 8.4 to build the qdq onnx to engine on orin, is this correct?

hello, @i-riyad I check my preprocess code, is ok. and i manually disabled some layers, F1 score droped to 6%. Do we have method to analysis sensitive layer for onnx model?

**Comment by i-riyad** [comment]
> Do we have method to analysis sensitive layer for onnx model?

You can try [ORT debugging ](https://github.com/microsoft/onnxruntime-inference-examples/tree/main/quantization/image_classification/cpu#debugging)tool. We will look into supporting yolo v8.
